### PR TITLE
GCS Server add SIGTERM signal handler

### PR DESCRIPTION
**Pull request by ffbin**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_server_main.cc [hunk 1]
```diff
@@ -62,6 +64,27 @@ int main(int argc, char *argv[]) {
   gcs_server_config.redis_port = redis_port;
   gcs_server_config.redis_password = redis_password;
   gcs_server_config.retry_redis = retry_redis;
-  ray::gcs::GcsServer gcs_server(gcs_server_config);
+  ray::gcs::GcsServer gcs_server(gcs_server_config, main_service);
+
+  // Destroy the GCS server on a SIGTERM. The pointer to main_service is
+  // guaranteed to be valid since this function will run the event loop
+  // instead of returning immediately.
+  // We should stop the service and remove the local socket file.
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_server_main.cc:72]:
```suggestion
```
GCS server doesn't have "local socket file"

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 2]
```diff
@@ -112,24 +114,17 @@ void GcsServer::Start() {
...
-
-  is_stopped_ = true;
-  RAY_LOG(INFO) << "Finished stopping gcs server.";
+  if (!is_stopped_) {
+    RAY_LOG(INFO) << "Stopping gcs server.";
+    // Shutdown the rpc server
+    rpc_server_.Shutdown();
+
+    is_stopped_ = true;
+    RAY_LOG(INFO) << "Finished stopping gcs server.";
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_server.cc:126]:
```suggestion
    RAY_LOG(INFO) << "GCS server stopped.";
```

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 3]
```diff
@@ -112,24 +114,17 @@ void GcsServer::Start() {
...
-  // Shutdown the rpc server
-  rpc_server_.Shutdown();
-
-  // Stop the event loop.
-  main_service_.stop();
-
-  is_stopped_ = true;
-  RAY_LOG(INFO) << "Finished stopping gcs server.";
+  if (!is_stopped_) {
+    RAY_LOG(INFO) << "Stopping gcs server.";
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_server.cc:121]:
```suggestion
    RAY_LOG(INFO) << "Stopping GCS server.";
```

File: src/ray/gcs/gcs_server/gcs_server_main.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server_main.cc]: fixed

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 2 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.cc]: fixed

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 3 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.cc]: fixed

**Review by raulchen — APPROVED** [review]

**Pull request merged**